bungie_oauth.py

```diff
@@ -0,0 +1,138 @@
+import base64
+import shelve
+import datetime
+import requests
+
+from dataclasses import dataclass
+from typing import Dict, Iterable, Set
+
+
+@dataclass
+class BungieOauthToken:
+    access_token: str
+    access_expires_at: datetime.datetime
+    refresh_token: str
+    refresh_expires_at: datetime.datetime
+
+
+class BungieOauth:
+    api_root = "https://www.bungie.net/Platform"
+    token_uri = f"{api_root}/app/oauth/token/"
+
+    def __init__(self, api_key, client_id, client_secret):
+        self.api_key = api_key
+        self.client_id = client_id
+        self.client_secret = client_secret
+
+        auth = base64.b64encode(f"{client_id}:{client_secret}".encode("ascii"))
```

Comment:
Minor question for my understanding: Is there a reason we encode ascii just to decode a few lines down?

Reply:
Good question. It's all down to the way Python handles strings and bytes.
The encode is because b64encode expects a byte sequence, not a string, hence the encode.
The later decode is to take the returned byte sequence, and converts it to a string for use in the string interpolation for the header.

I could just call encode/decode and use the default `utf-8`, but the range of characters that Bungie is using is only ascii (and the example I found was already using ascii - https://stackabuse.com/encoding-and-decoding-base64-strings-in-python/#encodingstringswithpython)

https://stackoverflow.com/questions/8908287/why-do-i-need-b-to-encode-a-string-with-base64/41437531 goes into more depth